Provide group key derivation at genesis proof

halseth opened the pull request:
This PR adds changes the group key derivation to rely on a `asset_raw_key` and `asset_script_root` provided in the asset proof at genesis. This lets the verifier derive and verify the `asset_group_key`.

TODO:
- [x] Require `raw_key` to be provided with proof at genesis.
- [x] Verify minting signature against group key.
- [x] Provide valid witness against group key when minting.

~~It also renames all occurrences of `asset_family_key` to `asset_group_key`.~~ Builds on #35 

See https://github.com/lightninglabs/taro/issues/91

halseth committed:
bip-tap: make minting witness encode asset internal key

This let the verifier check that the `asset_group_key` is derived
correctly from the internal key, and that the signature is valid.

halseth commented:
A question that comes to mind is 

How is the receiver of an asset able to derive the group key if the asset in question comes from a second tranche of issued assets?

Since the group key is derived from the original `asset_id`, just knowing the minting genesis of the received asset is not enough.

Roasbeef commented:
> How is the receiver of an asset able to derive the group key if the asset in question comes from a second tranche of issued assets?

Why do they need to derive it? Before they receive an asset, they'd first verify the base universe issuance proof of the OG asset. Once they have that, then they have access to the raw group key, and the tweaked version that's used everywhere else. 

Example flow:
  * Realize they like the latest Cricket player cards
  * Go to import the latest proof so they know how many they are
  * Realize that references a group key, so they should import that one first (obtain the raw key) 
  * Import the new one.

Roasbeef commented on bip-taro.mediawiki:
```diff
@@ -205,7 +205,7 @@ The Taro asset tree is a core component of the system. An asset tree is an
 MS-SMT commitment to a series of asset identifiers, grouped by their genesis
 asset ID. A given Taro asset tree is composed of two nested MS-SMT instances:
 
-# The first level maps an <code>asset_id</code> or <code>asset_key_family</code> to a sub-tree root hash of a given asset.
```
Thought we had already done this? 🤔 cc @jharveyb

Roasbeef replied:
Back then nope, but should be able to rebase and take away these three commits now 🫡

Roasbeef commented on bip-taro.mediawiki:
```diff
@@ -534,11 +534,14 @@ where:
 ** <code>asset_witness</code>: is a serialized witness in an identical format as Bitcoin's Segwit witness field. This field can only be blank if <code>prev_asset_id</code> is blank.
+*** If <code>prev_asset_id</code> is all zeroes, the witness will be either
+**** empty, to indicate this asset does not support further issuance. In this case <code>asset_group_key</code> MUST NOT be present.
+**** on the format <code><asset_key_internal> <signature></code>, where <code>asset_key_internal</code> is the public key used to derive the <code>asset_group_key</code>, and <code>signature</code> is a 64-byte BIP 340 signature over the <code>asset_id</code> for <code>asset_group_key</code>. In this case <code>asset_group_key</code> MUST be present.
```
"on the format" 

Seems like something in front of this got chopped off?

halseth replied:
It's a bullet list prepended with "either":

<img width="946" alt="Screenshot 2023-01-10 at 11 06 09" src="https://user-images.githubusercontent.com/3765732/211522107-1ddfbb9c-4f3a-4685-9b67-356f1636eae3.png">

Roasbeef reviewed (changes requested):
Looks directionally correct at a high level, however we need some other changes:
  * Update to VM validation w.r.t how to handle an asset that has the `prev_asset_id` be all zeroes:
      * Need to specify what the sighash in this case should be
      * If we want to make things mega flexible, then the witness here would be a control block of a single script path (can extend in the future) that's just the sig. We _could_ also change the derivation here to just stick the relevant info in the tapscript leaf (meta, tag, etc, etc). 

The extra step would then mean that the final group key has a tapscript root. Each time you meant, you need to satisfy one or the leaves (there can be many). This opens the door to having more specialized logic (multi-sig, etc), and can even add additional functionality via new op codes that isn't actually in script: zkp for w/e.

Roasbeef commented:
> Update to VM validation w.r.t how to handle an asset that has the prev_asset_id be all zeroes:

Also need to make explicit that for initial genesis, validation needs to:
  * Verify that the group key contained in the asset TLV matches up w/ what you'd derive given the raw group key in the witness (the internal group key) and the asset minting params (tag, meta, etc)
  * Then we can either just leave the thing being signed to be the asset ID, or use the same virtual tx structure (turn the asset TLV into a bitcoin tx) and run the tapscript VM on that. 

Also need to define group sig validation for future issuance events. In this case, when the witness is just a signature, then it's effectively the keyspend path (you've already verified the script path spend to reveal the internal key a.k.a the raw group key).

halseth commented:
> > How is the receiver of an asset able to derive the group key if the asset in question comes from a second tranche of issued assets?
> 
> Why do they need to derive it? Before they receive an asset, they'd first verify the base universe issuance proof of the OG asset. Once they have that, then they have access to the raw group key, and the tweaked version that's used everywhere else.
> 
> Example flow:
> 
> * Realize they like the latest Cricket player cards
> * Go to import the latest proof so they know how many they are
> * Realize that references a group key, so they should import that one first (obtain the raw key)
> * Import the new one.

I'm thinking of the very basic case, where an asset is issued without a base universe.

Initially I was thinking "if somebody sends me a random asset, how do I find the original asset id?", but this won't be a problem, since no one can send you a random asset without you importing it first (and creating a a receive address)?

halseth commented:
So here's a summary of what I think is a flexible approach (with some inline questions/TODOs):
* Change the group key to be `asset_group_key = asset_internal_key + TaggedHash(asset_internal_key, taproot)` where the `taproot` commits to what the current group key commits to (`genesis_outpoint || output_index || asset_type`) in addition to any scripts. 
  * This let us embed scripts that can gate minting in addition the the regular key spend.
  * TODO: can't we just simply commit to the original genesis `asset_id` here?
* When minting a second tranche of assets, the `asset_witness` would present a valid spend of the group key, either using the key spend path, or satisfying one of the tap leafs. I think it's a good idea to also require this for the original mint.
* When minting the original asset (genesis tranche), one must somehow prove that the group key commits to the above data. I suggest we add this data to another TLV `asset_group_control_block = <commitment control block>` at genesis, such that one can derive the full group key without revealing the scripts. This also makes it easy to distinguish a genesis mint from a second tranche mint.
  * TODO: To avoid the possibility for committing to more than one `asset_id` we must require this leaf to be in a deterministic position.
* We can use the virtual tx mapping to validate such a minting transaction, but we must be sure to sign some pseudo-random data from the Bitcoin transaction itself to avoid signature replay in a different transaction.

If this sounds like the direction you were thinking, I will revise the BIP changes and the implementation. @Roasbeef

Roasbeef commented:
> Change the group key to be

Shouldn't be asset internal key, but the "internal group key" which is now revealed in the genesis witness. 

>  commits to what the current group key commits to (genesis_outpoint || output_index || asset_type) in addition to any scripts.

Where would that commitment lie? In the tapscript tree itself? 

Re the other question, I think we want to either:
  1.  tweak internal key (single/double tweak from lnd) with the asset ID (since that commits to those other fields) 
  2.  use the normal tapscript root hash, then also commit to the asset ID in a leaf 
     * this would mean we'd also want a merkle branch into this as well 
   3. do something _slightly_ custom and commit to: `assetID || tapscriptRoot`

> When minting a second tranche of assets, the asset_witness would present a valid spend of the group key, either using the key spend path, or satisfying one of the tap leafs. I think it's a good idea to also require this for the original mint.

SGTM. The "message" being signed here, would then just be the virtual transaction we'd create that would be used to validate a normal spend. This commits to the entire execution environment. 


> When minting the original asset (genesis tranche), one must somehow prove that the group key commits to the above data

Control block sounds good there, since it forces a reveal of the internal key. Then the verifiers would double check the tweak mapping. Then see question above about how we want to bind the internal key or asset ID information. 

> We can use the virtual tx mapping to validate such a minting transaction, but we must be sure to sign some pseudo-random data from the Bitcoin transaction itself to avoid signature replay in a different transaction.

SGTM. Can you elaborate a bit w.r.t what you see as the replay risk here? The virtual transaction ends up binding to the _Bitcoin_ previous outpoint (assuming we have "genesis" be an actual bitcoin prev out, and then blank prev asset ID fields).

halseth commented:
After some offline discussion, we've decided to tweak the raw key twice:
```
internalGroup = singleTweak(rawGroupKey, assetID)
groupKey = tapTweak(internalGroup, tapScriptRoot)
```
That way we'll commit to the genesis asset id and a script tree that will make it flexible for later.

To verify this group key, the `rawGroupKey` must be revealed at asset genesis. Instead of setting this in the witness, we'll add it to a new TLV field that is only present in the genesis tx. (This has the added benefit that it makes it trivial to distinguish a genesis mint from a second tranche mint).

Roasbeef commented:
@halseth what about making that key a _required_ field in the proof file format similar to what we did with https://github.com/Roasbeef/bips/pull/34? In that PR, we made the meta a hash, then force reveal of that in the proof format for the newly minted assets. 

Then that way, it isn't adding a new special case in the asset TLV itself. Also proof verification is where this actually matters in the first place as well.

halseth commented:
(The "key family" -> "group key" rename I split off into a new PR to reduce the diff: #35) 

> @halseth what about making that key a _required_ field in the proof file format similar to what we did with #34? In that PR, we made the meta a hash, then force reveal of that in the proof format for the newly minted assets.
> 
> Then that way, it isn't adding a new special case in the asset TLV itself. Also proof verification is where this actually matters in the first place as well.

I moved the raw key and script root needed to derive the group key to the proof file - much better indeed! Nice that we are able to almost do no changes to the asset TLV, as the group key derivation is only needed when verifying the proof 👍 

I also changed the group key definition to be built from `SingleTweak` and `TapTweak` functions - should we fully write them out here or link to the relevant BIPs?

Finally, I made the group key commit to the `asset_id` directly, as that one is already derived from the unique data.

What remains is providing a valid witness for the group key when minting. I will add that here and to the implementation if you agree with this group key derivation @Roasbeef

halseth committed:
bip-tap proof file: also first transition now requires witness validation

halseth committed:
bip-tap vm: specify that minting transition witnesses must be verified

halseth commented:
Updated with verification of witness during VM state transition for minting.

halseth commented:
Changed to having an all-zero `asset_script_key` indicate a minting transction (instead of all-zero `prev_asset_id`) since this introduces the `prev_outpoint` as a source of randomness into the virtual tx, making sure the signature for the group key cannot be re-used.

Roasbeef commented:
> I moved the raw key and script root needed to derive the group key to the proof file - much better indeed! Nice that we are able to almost do no changes to the asset TLV, as the group key derivation is only needed when verifying the proof 👍

👍 

I think for us to remove the old field, we'll also need to do a version bump now that we're "officially" on testnet. 

> I also changed the group key definition to be built from SingleTweak and TapTweak functions - should we fully write them out here or link to the relevant BIPs?


I think it makes sense to write them out in full. 

> Changed to having an all-zero asset_script_key indicate a minting transction (instead of all-zero prev_asset_id) since this introduces the prev_outpoint as a source of randomness into the virtual tx, making sure the signature for the group key cannot be re-used.

Can you elaborate on this? One can already consider the asset ID as that source of randomness since it already includes the prev out iiuc. 


How could the signature be re-used other than somehow colliding an asset ID?

Roasbeef commented on bip-taro.mediawiki:
```diff
@@ -392,18 +392,16 @@ derived using the first input outpoint as a sort of chain randomness to ensure
 
 where:
-* <code>asset_key_internal</code> is a 32-byte public key as defined by
+* <code>asset_internal_group_key = singleTweak(asset_raw_key, asset_id)</code> 
```
So we can either prescribe that the asset iD is verified, or "unroll" the computation like we did above. Leaning towards the latter. 

There's another related change here: rn we include all the "asset genesis" (so the input to the asset ID) within each leaf TLV. Instead we can just have the `asset_id` there (single field), and move the genesis reveal to the genesis proof like we've done with the asset meta reveal.

halseth replied:
Genesis reveal moved to proof file.

Roasbeef commented on bip-taro.mediawiki:
```diff
@@ -392,18 +392,16 @@ derived using the first input outpoint as a sort of chain randomness to ensure
-* <code>output_index</code> (4 byte, big-endian) is the index  of the output
-which contains the unique Taro commitment in the genesis transaction. 
-* <code>asset_type</code> (1 byte) is the type of the asset being minted.
+* <code>root<code> is a taproot that can hold arbitrary asset scripts enforcing
```
`group_script_root`?

Roasbeef replied:
One other thing I'd like us to consider here is the idea of also starting to timestamp the base/canonical universe information on chain. 

I've started to solidify the concept here: https://github.com/Roasbeef/bips/blob/bip-tap/bip-tap-universe.mediawiki#canonical-root-universe-state-transition-rules

At a high level, we'd:
  * Add a new field to opt into enforced chain commitments of the Universe root (total supply)
  * If set, then the first output in the minting transaction must use the asset group key (in the context of this diff, I think it'd be the `asset_raw_key`) as the internal key of that output. 
  * That output would then commit to the new universe root (modified commitment) that includes the issued asset

Then for each other asset:
  * You can only mint by spending that output created above (linked issuance)
  * The first output inherits the above, and also should be a new valid commitment 

So basically trying to re-use the group key, such that you also need to satisfy it in order to do further issuance. Then this way, ppl can watch a single output on chain to know when more assets are issued (run the numbers, etc, etc). 

The goal here is that we'd sort of _force them_ (ideally we also support update/rotation of the policy too) to use the exact same tapscript root as the tap layer on the Bitcoin layer. This means you need to be able to move the Bitcoin output to make a new tap asset in the group and the other way around. 

I think we can achieve this if we add some verification/restrictions on the tapscript sibling hash (so the thing to the left/right of the tap/universe commitment). 

Other variations exists too: 
  * We can derive a NUMS key from the OG group key, then use that to always force script reveal (that sibling hash is revealed as you go to the left/right of the tree where all the scripts live). 

I think it's still nice to also retain this new tap witness requirement on the asset layer as you can have other functionality not yet deployed on script (other sigs, zkps, etc, etc).

jharveyb replied:
Doesn't this only force use of the same internal key, but not the actual tapscript root & tree? So you could use different trees on the Bitcoin and Tap layers (not sure why you'd want to though).

Roasbeef replied:
> Doesn't this only force use of the same internal key, but not the actual tapscript root & tree? So you could use different trees on the Bitcoin and Tap layers (not sure why you'd want to though).



Correct that it's just the internal key that's being enforced. Agree you'd also want to have it be mirrored as much as possible, which would then also require a proof of tapscript sub-tree equivalence.

Roasbeef commented on bip-taro.mediawiki:
```diff
@@ -532,17 +530,20 @@ where:
 ** <code>prev_asset_id</code>: references the previous asset input by the input position within the transaction, the asset ID of the previous asset tree, and the asset script hash.
 *** This value is included in any signatures generated within the asset witness itself, serving a purpose similar to the previous outpoint in vanilla Bitcoin.
-*** If this field is all zeroes, then that indicates a ''new'' asset is being created.
+*** If the <code>asset_script_key</code> field is all zeroes, then that indicates a ''new'' asset is being created.
```
Need to also tie this to an asset TLV version change if we decide it's actually necessary. 

Also how can the script key be all zeroes when we need to sign with that (or reveal a tapscript path) to be able to move it again?

halseth replied:
This refers to the _prev_ `asset_script_key`, from `prev_asset_id = prev_outpoint || asset_id || asset_script_key`.

Renamed it to `prev_asset_script_key` for clarity: https://github.com/Roasbeef/bips/pull/33/commits/57ee255ac988667f84a531687f039d3d6fa86d86

Roasbeef commented on bip-taro.mediawiki:
```diff
@@ -532,17 +530,20 @@ where:
+** <code>asset_witness</code>: is a serialized witness in an identical format as Bitcoin's Segwit witness field. 
+*** If <code>asset_script_key</code> is all zeroes, the witness will be either
+**** empty, to indicate this asset does not support further issuance. In this case <code>asset_group_key</code> MUST NOT be present.
+**** non-empty, in case the witness will be interpreted according to BIP-341 validtation rules on the <code>asset_group_key<code>. In this case <asset_group_key> MUST be present.
```
Link to section in the VM BIP about how to verify it?

halseth replied:
Done: 7f5b921c9a58e96f7b4ab32958be43927cbeae3e

Roasbeef commented on bip-taro-proof-file.mediawiki:
```diff
@@ -124,6 +124,9 @@ blob with the following format:
 *** [<code>...*byte</code>:<code>nested_proof_map</code>]
+* type: 10 (<code>asset_group_key_reveal</code>)
+** value: 
+*** [<code>64*byte</code>:<code>pub_key || asset_script_root</code>]
```
I think only the raw public key is needed? Since with that you can get to the internal key, then you'd force (?) an initial script path reveal so the internal key is known.

halseth replied:
The idea here was that if keypath spend is used with the group key, then the taproot is not revealed in the witness (only a signature), which is why we add it here. Otherwise I don't think one could actually verify the group key derivation in that case?

halseth committed:
bip tap: move asset_id reveal to proof file

halseth committed:
bip-tap proof: reveal asset_group_key at genesis proof

halseth committed:
bip-tap: clarify prev_asset_id and prev_asset_script_key

To avoid confusing it with the "next" asset_script_key and asset_id, we
rename the fields slightly.

halseth commented:
> I think for us to remove the old field, we'll also need to do a version bump now that we're "officially" on testnet.

Is there a clean way to do this without a bunch of if-elses in the BIP? Personally find it easier to just make it a breaking change, but your call obviously.


> > Changed to having an all-zero asset_script_key indicate a minting transction (instead of all-zero prev_asset_id) since this introduces the prev_outpoint as a source of randomness into the virtual tx, making sure the signature for the group key cannot be re-used.
> 
> Can you elaborate on this? One can already consider the asset ID as that source of randomness since it already includes the prev out iiuc.
> 
> How could the signature be re-used other than somehow colliding an asset ID?

If `prev_asset_id` is set all-zeroes, the `asset_id` is not actually signed anymore. Changing this to `prev_asset_script_key` being all-zeroes, retains the signature over the `asset_id` and `prev_outpoint`:
`prev_asset_id = prev_outpoint || asset_id || 00...00`

Without this change an existing signature could be reused to mint more tranches of the same asset (since the  same signature would be valid for the group key in a different minting transaction and different `asset_id`).

halseth commented:
Rebased and addressed comments. @Roasbeef PTAL

Roasbeef commented:
> Is there a clean way to do this without a bunch of if-elses in the BIP? Personally find it easier to just make it a breaking change, but your call obviously.

I think it would be something like:
  * specify that the existing field are for tx version 1 
  * add a section of fields for tx version 2

Correct that we don't _have_ to introduce a new version, but imo it's a useful exercise at this stage of the game to make sure stuff like this works as expected. 

> If prev_asset_id is set all-zeroes, the asset_id is not actually signed anymore. 

I think we can ensure the `asset_id` is still signed if we modify the virtual genesis transaction accordingly. IIUC [with this commit](https://github.com/Roasbeef/bips/pull/33/commits/d96fb3b1307f39c8323b3f73a0c0fede29dce4d1), the virtual tx is constructed as:
  * Input is a smt root with the output leaf with the prev asset witness empty 
  * Output is an smt root with the group key as the output (this being the original group key revealed for the very first genesis)

The output leaf serialization still has the `asset_id` encoded in the TLV leaf. As a result, it is still signed, and we can reserve an all zero script key for other purposes.

jharveyb commented:
This version looks good, pending resolving the changes w.r.t. `prev_asset_id`.

Checking my understanding of the net changes from this PR:

- Change the asset leaf TLV where `asset.asset_genesis` is replaced with `asset,asset_id`, and the `asset_genesis` field is moved into the genesis proof as a mandatory field. (Receivers are guaranteed to have `asset_genesis`, but its no longer duplicated in all proof objects of an asset history).

- Change the asset group key derivation s.t. we commit to a tapscript root with an additional key tweak, and replace the first key tweak with `asset_id` instead of `H(internal_key || genesis_outpoint || outpoint_index || asset_type)`.

- Change the asset leaf TLV s.t. `asset.asset_group_key` only stores the tweaked group key, and the Schnorr sig verifying an issuance is now moved to the `prev_asset_witnesses` array. (That sig can now also be any witness like a script path reveal).

- Add a TLV field to the proof object format, the `asset_group_key_reveal`, that includes the internal key for the asset group key and the tapscript root used as the secondary key tweak for the asset group key. This maintains the ability for asset receivers to verify to tweaked group key.

And the user-facing impact being smaller proofs for ungrouped assets, and the ability to authorize asset reissuance with a Tapscript path reveal vs. only a signature.

halseth commented:
> I think we can ensure the `asset_id` is still signed if we modify the virtual genesis transaction accordingly. IIUC [with this commit](https://github.com/Roasbeef/bips/pull/33/commits/d96fb3b1307f39c8323b3f73a0c0fede29dce4d1), the virtual tx is constructed as:
> 
> * Input is a smt root with the output leaf with the prev asset witness empty
> * Output is an smt root with the group key as the output (this being the original group key revealed for the very first genesis)
> 
> The output leaf serialization still has the `asset_id` encoded in the TLV leaf. As a result, it is still signed, and we can reserve an all zero script key for other purposes.

You're right, using this construction it will be properly included in the signature:
1. The `asset_id` is derived from an input in the minting transaction (this is the uniqueness we need for it not to be replayable).
2. The `asset_id` is then committed to by the output in the minting transaction.
3. The output is then signed‚ using the virtual tx construction, by the group key.

This should properly ensure that the signature cannot be reused, since no other minting event can reuse the `asset_id`.

halseth commented:
Updated the PR to use version 1: https://github.com/Roasbeef/bips/pull/33/commits/65a3a6ae78575dd07519ddd58363e51ebd72de8e

Also went back to using a blank `prev_asset_id` (renamed to `prev_asset_input` now) to indicate minting, [as per this insight](https://github.com/Roasbeef/bips/pull/33#issuecomment-1572866392).

PTAL @Roasbeef @jharveyb

jharveyb commented on bip-tap.mediawiki:
```diff
@@ -521,34 +523,39 @@ An asset leaf is a serialized TLV blob, with the following key-value mappings:
 *** [<code>33*byte</code>:<code>pub_key</code>]
-* type: 10 (<code>asset_group_key</code>)
-** value: 
+* type: 10 (<code>asset_group_key</code>) [version = 0, deprecated]
```
Maybe better to spell out the V1 Asset TLV explicitly, since the type numbers should also be updated to follow the even/odd rule?

Roasbeef replied:
I think we'd need to first step back and agree on what even/odd in this context means. Is it something like:
  * odd is for things that just need to be serialized properly in order to validate sigs 
  * even is for things that need to be known in order to properly validate a state transition or other higher level protocol rules

jharveyb commented on bip-tap-vm.mediawiki:
```diff
@@ -73,13 +73,18 @@ splits are validated.
+## From asset version 1:
+### For a minting transaction, a copy of the output leaf with emptied <code>prev_asset_witnesses</code> is used, in addition to these modifications:
+#### If the minted asset has a group key, the <code>asset_script_hash</code> of the copied leaf should be set equal to the group key. This enforces that the state transition verification uses the group key when validating the spend.
+#### If the asset has no group key, the <code>asset_script_hash</code> field should be blank. This will short-circuit the state transition verification, allowing minting an asset that does not support emission.
```
I don't think this modification of `asset_script_hash` is actually necessary?

Right now we short circuit based on `HasGenesisWitness()`, with no change for grouped assets. I think for grouped asset minting we would want to NOT short circuit if the asset passes `HasGenesisWitness()` and also has a group key.

With that change, I don't think we can just swap in the group key later on in the VM vs. modifying the `asset_script_hash`.

jharveyb replied:
Unconditional short circuiting here:

https://github.com/lightninglabs/taproot-assets/blob/257ad39d53a04f1cf05d6c4abf30d972ede67da4/vm/vm.go#L286

jharveyb replied:
As discussed offline, would be good to be more explicit in describing that the actual asset leaf is not modified here, only the _copy_ made when performing actual virtual transaction validation.

It would also be good to spell out `is_valid_issuance_txn()` explicitly.

jharveyb replied:
As discussed offline, would be good to be more explicit in describing that the actual asset leaf is not modified here, only the _copy_ made when performing actual virtual transaction validation.

It would also be good to spell out `is_valid_issuance_txn()` explicitly.

Roasbeef replied:
Yeah I don't think we need to do anything different here. If this is a minting event w/o a group key, then we don't need to do anything.

jharveyb reviewed (commented):
Looks pretty good, just some final Qs about modifying the `asset_script_hash` and changes in VM short circuiting.

dstadulis commented:
Versioning will be added the more element in the implementation / spec and #33 is need (See https://github.com/lightninglabs/taproot-assets/issues/347) to be completed before then.

@halseth Do you think you'll be able to conclude this PR in the next two weeks?

Roasbeef commented on bip-tap-vm.mediawiki:
```diff
@@ -73,13 +73,13 @@ splits are validated.
 # Initialize a new empty MS-SMT tree as specified in [[./bip-tap-ms-smt.mediawiki|bip-tap-ms-smt]].
 # For each Taproot Asset input ''c_i'', identified in the <code>prev_asset_witnesses</code> field:
-## Serialize the referenced previous asset leaf (identified by <code>prev_outpoint || asset_id || asset_script_hash</code>) in TLV format.
+## Serialize the referenced previous asset leaf (identified by <code>prev_outpoint || prev_asset_id || prev_asset_script_hash</code>) in TLV format.
```
script_hash -> script_key ?

Roasbeef replied:
Looked to be matched below but not here.

halseth replied:
Yep, fixed: https://github.com/Roasbeef/bips/pull/33/commits/9241e0ae6af3b648ae05bc78eb66098ea49ceb0e

Roasbeef commented on bip-tap-proof-file.mediawiki:
```diff
@@ -129,6 +129,16 @@ blob with the following format:
+*** [<code>BigSize</code>:<code>metadata_len</code>]
+*** [<code>metadata_len*byte</code>:<code>metadata</code>]
+*** [<code>u32</code>:<code>output_index</code>]
+*** [<code>u8</code>:<code>type</code>]
```
Is this the asset type? Looks like we still repeat it below on the TLV serialization.

halseth replied:
Yes, this i the asset type. I think it is okay to repeat it here, since this is the proof file, and you need it to derive the asset_id

Roasbeef commented on bip-tap.mediawiki:
```diff
@@ -398,18 +398,17 @@ derived using the first input outpoint as a sort of chain randomness to ensure
-which contains the unique Taproot Asset commitment in the genesis transaction. 
-* <code>asset_type</code> (1 byte) is the type of the asset being minted.
+* <code>asset_internal_group_key = SingleTweak(asset_raw_key, asset_id)</code> 
+* <code>asset_raw_key</code> is a 32-byte public key as defined by [[../master/bip-0340.mediawiki|BIP-340]]
```
Move to be the first field defined?

Roasbeef replied:
also `asset_group_raw_key=? to further delineate

halseth replied:
Done: https://github.com/Roasbeef/bips/pull/33/commits/7d5d456e4292c01e8c01b9f6a30d82bf633ce7c4

Roasbeef commented on bip-tap.mediawiki:
```diff
@@ -398,18 +398,17 @@ derived using the first input outpoint as a sort of chain randomness to ensure
+* <code>asset_group_script_root</code> is a taproot that can hold arbitrary asset scripts enforcing conditions for minting the asset.
+
+and
+* <code>TapTweak</code> is the taproot tweak function as defined in [[../master/bip-0341.mediawiki|BIP-341]]. 
```
👍

Roasbeef commented on bip-tap.mediawiki:
```diff
@@ -398,18 +398,17 @@ derived using the first input outpoint as a sort of chain randomness to ensure
-* <code>output_index</code> (4 byte, big-endian) is the index  of the output
-which contains the unique Taproot Asset commitment in the genesis transaction. 
-* <code>asset_type</code> (1 byte) is the type of the asset being minted.
+* <code>asset_internal_group_key = SingleTweak(asset_raw_key, asset_id)</code> 
```
So just to clarify: this is just an _initial binding_ to the assetID that created the group key right? So then for future minting events, all is needed is that a valid witness over the virtual transaction is presented, and there's no new layer of tweaking needed to arrive at the key things _should_ be validated against?

halseth replied:
Yes, this binds the `asset_id` of the original minting transaction (the first tranche) to the group key. This tweak won't later change.

jharveyb replied:
Can't a 3rd party that is aware of the the final tweaked key and a following reissuance reuse a script witness from a previous mint?

Before we used the final tweaked key to produce a new signature, and the witness (signature) also committed to the new asset being reissued - AFAICT now we are able to provide a witness that doesn't commit to the new assetID, and can be valid without owning the private key for the final tweaked key.

What am I missing here?

halseth replied:
I think maybe the discussion here answers your question? https://github.com/Roasbeef/bips/pull/33#issuecomment-1600614525

jharveyb replied:
IIUC that comment describes the flow for reissuance via the 'key path' of the group key. What I'm thinking about is the 'script path'.

In that case, if (and only if) the script itself does not require a signature, then we would not be signing over the virtual TX, and the script could be replayed.

An example would be an asset group key that uses a NUMS key as the internal key, and one of the scripts in the committed tapscript tree is a simple hashlock with no signature requirement.

I think this is addressed by just not having scripts that do not also require signatures, which is similar to existing best practices for on-chain scripting. IMO this would be worth noting somewhere since the consequences are different here compared to on-chain UTXOs or even tap assets (where you would also need to control the anchor output to take advantage of script replay).

halseth replied:
That is true, good catch!

But yeah, I think this will be similar to a normal bitcoin script with no signature. The difference is that we here expect to be re-using the script tree as many time as we like, while a Bitcoin UTXO can be spent only once.

I don't know if there are any real solutions to it though. Even if we had the witness commit to the `asset_id`, the script path witness could be re-used/re-created/replayed for another mint just by changing the asset_id (since no private key is required, anyone could do this). 

Am I understanding it correctly now?

(and yeah, worth mentioning somewhere in the docs)

Roasbeef commented on bip-tap.mediawiki:
```diff
@@ -535,12 +534,15 @@ where:
+** <code>asset_witness</code>: is a serialized witness in an identical format as Bitcoin's Segwit witness field. 
+*** If <code>prev_asset_input</code> is all zeroes, the witness will be either
+**** empty, to indicate this asset does not support further issuance. In this case <code>asset_group_key</code> MUST NOT be present.
+**** non-empty, in case the witness will be interpreted according to [[../master/bip-0340.mediawiki#specification|BIP-341]] validation rules using the <code>asset_group_key</code> in place of the <code>asset_script_hash</code>. In this case <code>asset_group_key</code> MUST be present.
```
`asset_script_hash` isn't used anymore in favor of `asset_script_key`

halseth replied:
Done: https://github.com/Roasbeef/bips/pull/33/commits/7d5d456e4292c01e8c01b9f6a30d82bf633ce7c4

Roasbeef commented on bip-tap.mediawiki:
```diff
@@ -535,12 +534,15 @@ where:
+** <code>asset_witness</code>: is a serialized witness in an identical format as Bitcoin's Segwit witness field. 
+*** If <code>prev_asset_input</code> is all zeroes, the witness will be either
+**** empty, to indicate this asset does not support further issuance. In this case <code>asset_group_key</code> MUST NOT be present.
+**** non-empty, in case the witness will be interpreted according to [[../master/bip-0340.mediawiki#specification|BIP-341]] validation rules using the <code>asset_group_key</code> in place of the <code>asset_script_hash</code>. In this case <code>asset_group_key</code> MUST be present.
```
Where is `asset_raw_key` revealed? Since we need that to be revealed to verify the two levels of mapping.

Roasbeef replied:
Also I think we need to distinguish from case of an initial mint (`asset_raw_key` being revealed for the first time), and subsequent ones (valid witness over `asset_group_key` and the virtual tx revealed).

halseth replied:
`asset_raw_group_key` is only revealed in the proof file (first mint), and is never needed again. 

After you verify the initial mint, you know that the `asset_group_key` is valid for the given `asset_id` and there is no need to distinguish initial mint from subsequent mints: 
They both need to have the virtual tx being signed with the `asset_group_key`. The only difference is the initial mint needs to prove the asset id matches the asset genesis.

Roasbeef commented on bip-tap-proof-file.mediawiki:
```diff
@@ -164,6 +167,11 @@ where:
 * <code>challenge_witness</code> is an optional asset witness over a well-defined asset state transition that proves ownership of the <code>script_key</code> the asset currently resides at.
 * <code>asset_genesis</code>: is a mandatory field for genesis assets. The <code>asset_genesis</code> represents the preimage of the <code>asset_id</code>.
 ** This field MUST only be present for genesis asset proofs.
+* <code>asset_group_key_reveal</code>: is a mandatory field (for genesis assets having a group key) that reveals the raw internal signing key and script root needed to derive the asset group key: <code>asset_raw_key = asset_group_key_reveal[0:32]</code>, <code>asset_group_script_root = asset_group_key_reveal[32:64]</code>.
```
👍

Roasbeef commented on bip-tap-vm.mediawiki:
```diff
@@ -80,6 +80,10 @@ follows:
 # For each Taproot Asset input ''c_i'', identified in the <code>prev_asset_witnesses</code> field:
 ## Serialize the referenced previous asset leaf (identified by <code>prev_outpoint || prev_asset_id || prev_asset_script_hash</code>) in TLV format.
+### For a minting transaction, a copy of the output leaf with emptied <code>prev_asset_witnesses</code> is used, in addition to these modifications:
+#### If the minted asset has a group key, the <code>asset_script_hash</code> of the copied leaf should be set equal to the group key. This enforces that the state transition verification uses the group key when validating the spend.
```
Same here re generally cleaning up `script_hash` vs `script_key` across the spec

Roasbeef replied:
> of the copied leaf should be set equal to the group key.

Hmm, I think we still want to commit to this `script_key` value so validation context is bound completely. Otherwise a sig can re re-used for each leaf item? 

In other words, we can have validate be based on `asset_group_key`, while serializing `asset_script_key` in the TLV leaf. This way both are still committed to, and we have a single branch that uses `asset_group_key` when this case is detected.

halseth replied:
The modification only affects the referenced _previous leaf_, that is used to construct the previous outpoint for the virtual tx. The output leaf (still committing to `asset_script_key`) remains the same.

Or did I misunderstand your comment?

dstadulis commented:
Tracking comment: @jharveyb will take over the issue that this PR was addressing

halseth commented:
[Changed](https://github.com/Roasbeef/bips/pull/33/commits/0c062c1a90ea0f8fdad826bbf13b01eee3d802fc) `SingleTweak` definition to be aligned with the one lnd defines: https://github.com/lightningnetwork/lnd/blob/bf5aab9d5246daede3f499af880a5d7be6c6ee88/input/script_utils.go#L1418

Roasbeef commented:
I think this needs a rebase!

Roasbeef commented on bip-tap-proof-file.mediawiki:
```diff
@@ -129,6 +128,19 @@ blob with the following format:
 * type: 10 (<code>challenge_witness</code>)
 ** value:
 *** [<code>...*byte</code>:<code>challenge_witness</code>]
+* type: 11 (<code>asset_genesis</code>) [version >= 1]
```
So I think that we actually just will have this be v0, so starting from the "ideal" version.

halseth replied:
Removed version bump commit.

jharveyb commented:
Reading this over again to grok the virtual TX format, and I think we don't need to build a virtual TX at all for minted assets with no group?

In that situation, there is no witness to verify, so I don't see why we would construct a virtual TX. Currently we have an early exit of the VM for genesis assets, and I think now we would maintain that, *unless* the genesis asset is part of a group (and therefore has a group witness). Then we would need another early branch to make the special virual TX for a grouped asset minting, and proceed as normal.

halseth commented:
> Reading this over again to grok the virtual TX format, and I think we don't need to build a virtual TX at all for minted assets with no group?
> 
> In that situation, there is no witness to verify, so I don't see why we would construct a virtual TX. Currently we have an early exit of the VM for genesis assets, and I think now we would maintain that, _unless_ the genesis asset is part of a group (and therefore has a group witness). 

Yes, exactly: https://github.com/Roasbeef/bips/pull/33/commits/6ba5ddc1d24f60fa03fa1ccd9f977727785b34a2#diff-eb581d58c55516fe1407829febfad2ddd93007d14d0b4caabf8a33d7858f379bR210

> Then we would need another early branch to make the special virual TX for a grouped asset minting, and proceed as normal.

Virtual tx creation can work as before, as long as one does the change indicated here before creating the virtual tx: https://github.com/Roasbeef/bips/pull/33/commits/6ba5ddc1d24f60fa03fa1ccd9f977727785b34a2#diff-eb581d58c55516fe1407829febfad2ddd93007d14d0b4caabf8a33d7858f379bR83-R86

jharveyb commented:
> Virtual tx creation can work as before, as long as one does the change indicated here before creating the virtual tx: [6ba5ddc#diff-eb581d58c55516fe1407829febfad2ddd93007d14d0b4caabf8a33d7858f379bR83-R86](https://github.com/Roasbeef/bips/commit/6ba5ddc1d24f60fa03fa1ccd9f977727785b34a2#diff-eb581d58c55516fe1407829febfad2ddd93007d14d0b4caabf8a33d7858f379bR83-R86)

So my point was that step 2.1.a.b. should never happen, since for that minting type we've already exited the VM. Will have a PR up soon though.

halseth commented:
> > Virtual tx creation can work as before, as long as one does the change indicated here before creating the virtual tx: [6ba5ddc#diff-eb581d58c55516fe1407829febfad2ddd93007d14d0b4caabf8a33d7858f379bR83-R86](https://github.com/Roasbeef/bips/commit/6ba5ddc1d24f60fa03fa1ccd9f977727785b34a2#diff-eb581d58c55516fe1407829febfad2ddd93007d14d0b4caabf8a33d7858f379bR83-R86)
> 
> So my point was that step 2.1.a.b. should never happen, since for that minting type we've already exited the VM. Will have a PR up soon though.

Ah, I see what you mean: since we are exiting early, we don't even come to the point where we are checking the virtual TX, hence we don't need to create it in the first place.

That being said, we need the VM to be able to tell that it should exit early. And I think setting the `asset_script_key` ro zeroes in the virtual tx is and okay way to do that (it is the same code anyway to build the virtual tx).

halseth commented:
Rebased.

Roasbeef commented:
> And I think setting the asset_script_key ro zeroes in the virtual tx is and okay way to do that (it is the same code anyway to build the virtual tx).

IIUC, you're talking about the case where you have issuance, without a group key. Why isn't it enough to just check that a group key isn't present, then exit there?

halseth commented:
> > And I think setting the asset_script_key ro zeroes in the virtual tx is and okay way to do that (it is the same code anyway to build the virtual tx).
> 
> IIUC, you're talking about the case where you have issuance, without a group key. Why isn't it enough to just check that a group key isn't present, then exit there?

~~Yes, but the VM doesn't have access to the group key (only the virtual tx). But one could just never invoke the VM if it's not checking anything anyway.~~
Looking at the code again, this was wrong.  Looks like one can indeed do what you suggested, since we have the new asset available. So maybe that's a better way 😄

Roasbeef commented:
Landing this as is, @jharveyb will follow up with the final tweaks.
